src/io/buf_writer.rs

```diff
@@ -0,0 +1,547 @@
+        }
+    }
+
+    /// Gets a reference to the underlying writer.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # #![feature(async_await)]
+    /// # #![allow(unused_mut)]
+    /// # fn main() -> std::io::Result<()> { async_std::task::block_on(async {
+    /// use async_std::io::BufWriter;
+    /// use async_std::net::TcpStream;
+    ///
+    /// let mut buffer = BufWriter::new(TcpStream::connect("127.0.0.1:34254").await?);
+    ///
+    /// // We can use reference just like buffer
+    /// let reference = buffer.get_ref();
+    /// #
+    /// # Ok(()) }) }
+    /// ```
+    pub fn get_ref(&self) -> &W {
+        &self.inner
+    }
+
+    /// Gets a mutable reference to the underlying writer.
+    ///
+    /// It is inadvisable to directly write to the underlying writer.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # #![feature(async_await)]
+    /// # fn main() -> std::io::Result<()> { async_std::task::block_on(async {
+    /// use async_std::io::BufWriter;
+    /// use async_std::net::TcpStream;
+    ///
+    /// let mut buffer = BufWriter::new(TcpStream::connect("127.0.0.1:34254").await?);
+    ///
+    /// // We can use reference just like buffer
+    /// let reference = buffer.get_mut();
+    /// #
+    /// # Ok(()) }) }
+    /// ```
+    pub fn get_mut(&mut self) -> &mut W {
+        &mut self.inner
+    }
+
+    //    pub fn get_pin_mut(self: Pin<&mut Self>) -> Pin<&mut W> {
+    //        self.inner()
+    //    }
+
+    /// Consumes BufWriter, returning the underlying writer
+    ///
+    /// This method will not write leftover data, it will be lost.
+    /// For method that will attempt to write before returning the writer see [`poll_into_inner`]
+    ///
+    /// [`poll_into_inner`]: #method.poll_into_inner
+    pub fn into_inner(self) -> W {
+        self.inner
+    }
+
+    //    pub fn poll_into_inner(self: Pin<&mut Self>, _cx: Context<'_>) -> Poll<io::Result<usize>> {
+    //        unimplemented!("poll into inner method")
+    //    }
+
+    /// Returns a reference to the internally buffered data.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// # #![feature(async_await)]
+    /// # fn main() -> std::io::Result<()> { async_std::task::block_on(async {
+    /// use async_std::io::BufWriter;
+    /// use async_std::net::TcpStream;
+    ///
+    /// let buf_writer = BufWriter::new(TcpStream::connect("127.0.0.1:34251").await?);
+    ///
+    /// // See how many bytes are currently buffered
+    /// let bytes_buffered = buf_writer.buffer().len();
+    /// #
+    /// # Ok(()) }) }
+    /// ```
+    pub fn buffer(&self) -> &[u8] {
+        &self.buf
+    }
+
+    /// Poll buffer flushing until completion
+    ///
+    /// This is used in types that wrap around BufWrite, one such example: [`LineWriter`]
+    ///
+    /// [`LineWriter`]: struct.LineWriter.html
+    pub fn poll_flush_buf(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
```

Comment:
Let's make this method private. I'm not against making it public, but would consider it after publishing version 1.0. For now, let's just match the standard library's API...

Reply:
done!